### PR TITLE
feat: row and column totals

### PR DESCRIPTION
**Pull request by Birkbjo**
https://dhis2.atlassian.net/browse/TECH-1156

This implements support for calculating the total of each row/column. This depends on the properties `showColumnTotal` and `showRowTotal` in the `section`-part of the metadata. This can be set up in the Maintenance app under the `Hamburger`-menu in the `Data Set`-list. 
Note that due do the location of these properties, only `section`-forms are supported.

I've moved `CategoryComboTable` to a separate folder along with the new files. Even though this feature is only supported for section-forms the implementation has to be done in this component, and thus I grouped the new files in this folder as well. They also build upon the `.css`-file.

### CODE REVIEW

File: src/data-workspace/category-combo-table/use-value-matrix.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+            values: true,
+            active: true,
+        },
+    })
+
+    const matrix = useMemo(() => {
+        return dataElements.map((de) =>
+            sortedCOCs.map((coc) => getIn(values, `${de.id}.${coc.id}`))
+        )
+    }, [dataElements, sortedCOCs, active])
```
Comment (Birkbjo) [inline src/data-workspace/category-combo-table/use-value-matrix.js:22]: If we would want this to update every time the value changes, we would switch out `active` with `values`. 
Currently this works so that it only updated when `active` is changed - and that is always changed `onBlur`.

File: src/data-workspace/category-combo-table/category-combo-table.js [hunk 2]
```diff
@@ -8,55 +8,63 @@ import {
...
         data,
         categoryCombo.id
     )
 
-    // each element is a combination of category-options for a particular column
-    // this results in lists of category-options in the same order as headers are rendered
-    // the result is a client side computation of categoryOption-combinations
-    const optionsIdLists = categories.map((cat) => cat.categoryOptions)
-    const computedCategoryOptions = cartesian(optionsIdLists)
+    const sortedCOCs = useMemo(() => {
```
Comment (Birkbjo) [inline src/data-workspace/category-combo-table/category-combo-table.js:39]: I did some refactoring here, because at first I rendered the total-cells in this component - and thus had to subscribe to values in this component - but that means re-rendering the entire table on every-keystroke. I abandoned that, but kept some of the refactoring.

I think we could look into also refactoring the table into `CategoryComboTableHeader` -component and move some of the logic out of this component.

File: src/data-workspace/category-combo-table/use-value-matrix.js (hunk 1 above)
Comment (Mohammer5) [reply src/data-workspace/category-combo-table/use-value-matrix.js:23]: Please add this to the comment, otherwise it's not clear from reading the code

**Review by Mohammer5 — COMMENTED** [review]
This is just a suggestion: When using `useMetadata`, I find it a lot easier to read code when using one of the following styles:

```js
const metadata = useMetadata() // use metadata.data instead of data
const { data: metadata } = useMetadata() // use metadata instead of data
```

---

While reviewing the changes in `category-combo-table.js` I realised that it's incredibly hard to understand what's happening, and even harder to understand why things are happening. 

Could you refactor that file and try to split things into functions and/or components as well as adding comments / explanations for why things are the way they are? The matter of this topic is complex and complicated in nature and I think that's why we need to put extra care into trying to make it as transparent as possible for someone who's reading the code but doesn't have a solid understanding of the process like you

File: src/data-workspace/category-combo-table/calculate-totals.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        ? matrix[0].map((v) => Number(v) || 0)
+        : matrix.reduce((totals, currRow) =>
+              totals.map((currTotal, i) => sum(currTotal, currRow[i]))
+          )
+
+const sum = (acc, curr) => {
+    const a = Number(curr) || 0
+    const b = Number(acc) || 0
+    return a + b
+}
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: Could do this inline as well. But I don't really understand why you would prefer to use a lodash-function when it's so simple?

I'm not sure I follow, notice that this function returns an array... It does not help to provide a default value here, and actually it's deliberately not done because then it uses the first value of the array.
It returns an array of totals. Because it calculates all the sum for all column**s**.

File: src/data-workspace/category-combo-table/category-combo-table.js [hunk 4]
```diff
@@ -8,55 +8,63 @@ import {
...
-            Computed: ${computedCategoryOptions.length}
+            Computed: ${sortedCOCs.length}
             Server: ${categoryCombo.categoryOptionCombos.length})`
         )
     }
     // Computes the span and repeats for each columns in a category-row.
     // Repeats are the number of times the options in a category needs to be rendered per category-row
-    let catColSpan = computedCategoryOptions.length
+    let catColSpan = sortedCOCs.length
     const rowToColumnsMap = categories.map((c) => {
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/category-combo-table.js:68]: Yeah I've thought about refactoring this to a header component, but thought maybe this would be best in it's own PR, since it's not really relevant to this feature. The reason I started some refactoring was because I implemented this feature differently and thus touched upon this code, but figured it was not needed now.

There are comments here about what it does though. I guess it would be more clear in a separate component/function. This basically just computes the headers in the table, which is a bit complex due to the dynamic nature of the table, since you need to calculate the columnSpan and number of times each option should be repeated. This is explained in the comments, and I'm not sure this would be much clearer with more comments to be honest, but I can try in a refactor. 

The reason `columnSpan` is overridden is because it depends on the previous. Of course you could probably write this using a `previousColumnSpan` and calculate `currentColumnSpan` from this, but the end result is the same (with more lines). Would that make it clearer?

 We are creating a "nested"/pivot-table, and for it to make sense, all the options in the previous category need to "span" all options in the current category. `sortedCOCs.length` will refer to the number of rendered columns, and this is used as the base, because in the end the last category should have the same amount of columns - and reach span "1". But to get there, we are basically doing factorization in each step, and the result in each step is the span for that category.

Eg. table with one category is simple, each column has a span of 1, so this loop only happens once. Say there are 4 options; `sortedCOCs.length === 4, nrOfOptions ===4` `4/4 === 1`, span is 1. 

Eg. table with 2 categories. 
`Category1 : Location:     <Fixed, Outreach>`
`Category2: Age:   <1, >1.`
sortedCOCS.length === 4. 
`1. iteration (Location):
     catColSpan = 4/numberOfOptions === 4/2 === 2
`
`2. iteration (Age)
   catColSpan = 2/numberOfOptions === 2/2 === 1
`
You see that current columnSpan depends on previous. There are probably dozens of way to write this as well, and as mentioned maybe a `previousColumnSpan` is clearer?
I think that in situations like this, if you really want to understand the code you need to take some time to study the problem anyway.

Again, this might be more easier to follow in a `CategoryComboTableHeader`-component. But I do think we should keep that refactor out of this PR.

File: src/data-workspace/category-combo-table/category-combo-table.js [hunk 5]
```diff
@@ -85,13 +92,6 @@ export const CategoryComboTable = ({
         return c
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/category-combo-table.js:92]: Will tackle this in a refactor.

File: src/data-workspace/category-combo-table/calculate-totals.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: I agree that providing an initial value of `0` to `reduce` would be better, because then `sum` could be simplified to this.
```js
const sum = (acc, curr) => acc + (Number(curr) || 0)
```
(at which point you could even consider just using an anonymous arrow function instead of a named const.

I don't think we need to add another dependency to the project for this though.

[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: This would need a change to the reduce-function though? Notice how I'm taking advantage of how `reduce` uses the first value of the array if not default is provided. And the first value is an array... `totals` is not a number! It's an array.

File: src/data-workspace/category-combo-table/category-combo-table.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/data-workspace/category-combo-table/category-combo-table.js:68]: I'm also having a tough time completely following this logic. Quite likely that's because I am unfamiliar with some intricacies of what we are trying to do here. In any case, I think that doing some refactoring here would be very beneficial and we can look into simplifying things when we do that. I fully agree that it should be done in a subsequent PR.

**Review by HendrikThePendric — COMMENTED** [review]
The changes in here look good to me. I agree with @Mohammer5 that the file `src/data-workspace/category-combo-table/category-combo-table.js` is quite difficult to follow and would benefit from refactoring. I think it makes sense to do this in a separate PR though.

File: src/data-workspace/category-combo-table/calculate-totals.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: > Notice how I'm taking advantage of how `reduce` uses the first value of the array as the accumulator if not default is provided.

No I had not noticed that. I personally always like to provide initial values to reduce functions, but I guess there is nothing wrong with doing things this way.

Beyond that I still don't really understand why we need this conditional logic. When we talk about a matrix in this context, I am assuming we are simply talking about an array, which contains a bunch of items, which are also arrays with an equal length. So something like this:
```js
[
    [1,2,3,5],
    [1,2,3,5],
    [1,2,3,5],
]
```

And if that's right then the calculations are like also pretty basic:
```js
const rowTotals = (matrix, row = 0) => matrix[row].reduce((acc, currVal) => acc + (Number(currVal) || 0), 0)
const colTotals = (matrix, column = 0) => matrix.reduce((acc, currRow) => acc + (Number(currRow[column]) || 0), 0)
```

What is more complex about what we are doing?

File: src/data-workspace/category-combo-table/calculate-totals.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+export const calculateColumnTotals = (matrix) =>
+    matrix.length < 2
+        ? matrix[0].map((v) => Number(v) || 0)
+        : matrix.reduce((totals, currRow) =>
+              totals.map((currTotal, i) => sum(currTotal, currRow[i]))
+          )
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/calculate-totals.js:6]: Nop, again, note that when `reduce` does not have a default value, the first element in the array is the `accumulator`, the `second` will be the `current`-element. So `reduce` will through in case of `matrix.length === 1`.

File: src/data-workspace/category-combo-table/calculate-totals.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: Sorry, I missed the fact that you were not passing a column index to the `calculateColumnTotals` function. So you want the output to be an array. This also made me understand why you want to not provide an initial value, and let the function use the first matrix row instead. But I still don't see the need for conditional logic. The code below should work:
```js
const calculateRowTotals = (matrix, row = 0) =>
    matrix[row].reduce((total, currVal) => total + (Number(currVal) || 0), 0)

const calculateColumnTotals = (matrix) => matrix.reduce((totals, row) => {
    row.forEach((currVal, index) => {
        totals[index] = totals[index] + (Number(currVal) || 0)
    })
    return totals
})
```

[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: > `const colTotals = (matrix, column = 0) => 
    matrix.reduce((acc, currRow) => acc + (Number(currRow[column]) || 0), 0)`

Here you are changing the function though. The way the `ColumnTotals`-component works is that it renders the totals for all columns, and thus this function calculates all totals in one go. For `RowTotal` this is different, because they are rendered in-line for each row - and because you are already mapping over rows (when rendering the table), it's easier for that function to just calculate per-row.
In your function you would need an additional map to map over the columns. Then you would end up with something like

```javascript
const calculateColumnTotals = (matrix) => matrix[0].map((_c, i) => colTotals(matrix, i))
```
Which is fine, but I don't really see how that's so much better than the current solution. Guess it might be easier to read, and then you wouldn't need the conditional to check for matrix.length of one.

[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: Hmm... Seems like you are right, but I did end up in a situation where I got some error that resulted in the conditional. But it might've been that the matrix was completely empty? In that case we can just remove the conditional in the original function. Because then it works there as well, since the problem would be the outer reduce, not the inner map. 

And if I'm not mistaken your solution basically just changes the functional `.map` to an imperative `forEach`, the result should be identical.

[the remark this answers is not in the document]
Comment (HendrikThePendric) [reply src/data-workspace/category-combo-table/calculate-totals.js:12]: I think you are correct. And agree we should add some tests for these edge case, as you suggested on Slack.

**Review by HendrikThePendric — APPROVED** [review]
As far as I am concerned, this is good to go in as-is. Let's refactor the parts of the code that are hard to follow and were discussed in another PR.
